File: leetcode/problems/language-specific/go/23.merge-k-sorted-lists.go
/*
 * @lc app=leetcode id=23 lang=golang
 *
 * [23] Merge k Sorted Lists
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */

// * Divide and Conquer Solution | O(nlog(k)) Time | O(1) Space

func mergeKLists(lists []*ListNode) *ListNode {
	if len(lists) <= 0 {
		return nil
	}

	interval := 1
	for interval < len(lists) {
		for i := 0; i+interval < len(lists); i = i + (interval * 2) {
			lists[i] = mergeLists(lists[i], lists[i+interval])
		}

		interval *= 2
	}

	return lists[0]
}

func mergeLists(list1, list2 *ListNode) *ListNode {
	dummyHead := &ListNode{0, nil}
	cur := dummyHead
	for list1 != nil && list2 != nil {
		if list1.Val <= list2.Val {
			cur.Next = list1
			list1 = list1.Next
		} else {
			cur.Next = list2
			list2 = list2.Next
		}

		cur = cur.Next
	}

	if list1 != nil {
		cur.Next = list1
	} else {
		cur.Next = list2
	}

	return dummyHead.Next
}

// @lc code=end

File: leetcode/problems/language-specific/python/23.merge-k-sorted-lists.py
#
# @lc app=leetcode id=23 lang=python3
#
# [23] Merge k Sorted Lists
#

# @lc code=start

# * Divide and Conquer Solution | O(nlogk) Time | O(1) Space
# * k -> The number of linked lists | n -> The number of nodes in two lists while merging


# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next


class Solution:
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        if not lists or len(lists) <= 0:
            return None

        lists_len = len(lists)
        interval = 1
        while interval < lists_len:
            for i in range(0, lists_len - interval, interval * 2):
                lists[i] = self.mergeTwoLists(lists[i], lists[i + interval])
            interval *= 2

        return lists[0]

    def mergeTwoLists(
        self, list1: Optional[ListNode], list2: Optional[ListNode]
    ) -> Optional[ListNode]:
        if not list1 and not list2:
            return list1

        dummy_head = ListNode()
        cur = dummy_head
        while list1 and list2:
            if list1.val <= list2.val:
                cur.next = list1
                list1 = list1.next
            else:
                cur.next = list2
                list2 = list2.next
            cur = cur.next

        if list1:
            cur.next = list1
        elif list2:
            cur.next = list2

        return dummy_head.next


# @lc code=end
